azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -223,73 +223,141 @@ def _handle_pipeline_event(self, event):
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _do_connect(self, op):
+        """
+        Start connecting the transport in response to some operation
+        """
+        # function that gets called after we're connected.
+        @pipeline_thread.runs_on_pipeline_thread
+        def on_connected(op_connect):
+            logger.info(
+                "{}({}): connection is complete: {}".format(self.name, op.name, op_connect.error)
+            )
+            if op_connect.error:
+                logger.info(
+                    "{}({}): completing with failure because of connection failure".format(
+                        self.name, op.name
+                    )
+                )
+                op.error = op_connect.error
+                operation_flow.complete_op(stage=self, op=op)
+            else:
+                logger.info(
+                    "{}({}): connection is complete.  Continuing with op".format(self.name, op.name)
+                )
+                operation_flow.pass_op_to_next_stage(stage=self, op=op)
+
+        # call down to the next stage to connect.
+        logger.info("{}({}): calling down with Connect operation".format(self.name, op.name))
+        operation_flow.pass_op_to_next_stage(
+            self, pipeline_ops_base.ConnectOperation(callback=on_connected)
+        )
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def on_connected(self):
+        self.connected = True
+        PipelineStage.on_connected(self)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def on_disconnected(self):
+        self.connected = False
+        PipelineStage.on_disconnected(self)
+
+
+class SerializeConnectOpsStage(PipelineStage):
+    """
+    This stage is responsible for serializing connect, disconnect, and reconnect ops on
+    the pipeline, such that only a single one of these ops can go past this stage at a
+    time.  This way, we don't have to worry about cases like "what happens if we try to
+    disconnect if we're in the middle of reconnecting."  This stage will wait for the
+    reconnect to complete before letting the disconnect past.
+    """
+
+    def __init__(self):
+        super(SerializeConnectOpsStage, self).__init__()
         self.queue = queue.Queue()
         self.blocked = False
+        self.connected = False
 
     @pipeline_thread.runs_on_pipeline_thread
     def _execute_op(self, op):
-        # If this stage is currently blocked (because we're waiting for a connection
-        # to complete, we queue up all operations until after the connect completes.
+        # If this stage is currently blocked (because we're waiting for a connection, etc,
+        # to complete), we queue up all operations until after the connect completes.
         if self.blocked:
             logger.info(
-                "{}({}): pipeline is blocked waiting for connect.  queueing.".format(
+                "{}({}): pipeline is blocked waiting for connect/disconnect/reconnect.  queueing.".format(
```

Comment:
>waiting for connect [](start = 45, length = 19)

`waiting for a previous connect/disconnect/reconnect` may sound better.
Is there any possibility of knowing what was the exact previous operation ?

Reply:
Yeah, something like "pipeline is blocked waiting for a prior connect/disconnect/reconnect to complete"

---
In reply to: [311696139](https://github.com/Azure/azure-iot-sdk-python-preview/pull/160#discussion_r311696139) [](ancestors = 311696139)

Reply:
changed.  I could add code to keep track of the current operation, but I don't want to do that just for logging because then I need to make sure it's always correct which means a whole bunch of tests and I just don't want to go there.  

---
In reply to: [311702297](https://github.com/Azure/azure-iot-sdk-python-preview/pull/160#discussion_r311702297) [](ancestors = 311702297,311696139)